File: py/pe14.py
#! /usr/bin/env python3

from sys import argv

cache = {1:1}

def chain(n):
    if not cache.get(n, 0):
        if n & 1:
            cache[n] = 1 + chain(3 * n + 1)
        else:
            cache[n] = 1 + chain(n >> 1)
    return cache[n]

def pe14(limit):
    """
    Which starting number produces the longest Collatz chain?
    """
    m, mi = 0, 0
    for i in range(3, limit, 2):
        c = chain(i)
        if m < c:
            m, mi = c, i
    print(mi, m)

def main():
    argc = len(argv)
    if argc > 1:
        try:
            n = int(argv[1])
        except ValueError:
            print("  python pe14.py [integer]")
            return
    else:
        n = 1000000
    
    pe14(n)

if __name__ == "__main__":
    main()

File: c/src/pe14.c
#include <pe14.h>

#include <inttypes.h>
#include <stdio.h>

uint32_t count(uint32_t n) {
  uint32_t c = 1;
  while (n > 1) {
    if (n & 1) {
      n = 3 * n + 1;
    } else {
      n >>= 1;
    }
    ++c;
  }

  return c;
}

void pe14(uint32_t begin, uint32_t limit) {
  uint32_t m = 0;
  uint32_t mi = 0;
  if ((begin & 1) == 0) {
    uint32_t c = count(begin);
    if (m < c) {
      m = c;
      mi = begin;
    }
    ++begin;
  }

  for (uint32_t i = begin; i < limit; i += 2) {
    uint32_t c = count(i);
    if (m < c) {
      m = c;
      mi = i;
    }
  }

  printf("%" PRIu32 " produces %" PRIu32 " Collatz chain\n", mi, m);
}

int pe14_main(void) {
  pe14(500001u, 1000000u);

  return 0;
}
